nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/SplitText.java

```diff
@@ -230,100 +270,112 @@ public void onTrigger(final ProcessContext context, final ProcessSession session
-                        errorMessage.set("Header Line Count is set to " + headerCount + " but file had only " + headerLinesCopied + " lines");
+                    final SplitInfo headerInfo = readHeader(headerCount, headerMarker, in, headerStream, true);
+                    if (headerInfo.lengthLines < headerCount) {
+                        errorMessage.set("Header Line Count is set to " + headerCount + " but file had only "
+                                + headerInfo.lengthLines + " lines");
                         return;
                     }
 
                     while (true) {
-                        if (headerCount > 0) {
-                            // if we have header lines, create a new FlowFile, copy the header lines to that file,
-                            // and then start copying lines
-                            final IntegerHolder linesCopied = new IntegerHolder(0);
-                            FlowFile splitFile = session.create(flowFile);
-                            try {
-                                splitFile = session.write(splitFile, new OutputStreamCallback() {
-                                    @Override
-                                    public void process(final OutputStream rawOut) throws IOException {
-                                        try (final BufferedOutputStream out = new BufferedOutputStream(rawOut)) {
+                        FlowFile splitFile = session.create(flowFile);
+                        final SplitInfo flowFileInfo = new SplitInfo();
+
+                        // if we have header lines, write them out
+                        // and then start copying lines
+                        try {
+                            splitFile = session.write(splitFile, new OutputStreamCallback() {
+                                @Override
+                                public void process(final OutputStream rawOut) throws IOException {
+                                    try (final BufferedOutputStream out = new BufferedOutputStream(rawOut)) {
+                                        long lineCount = 0;
+                                        long byteCount = 0;
+                                        // Process header
+                                        if (headerInfo.lengthLines > 0) {
+                                            flowFileInfo.lengthBytes = headerInfo.lengthBytes;
+                                            byteCount = headerInfo.lengthBytes;
                                             headerStream.writeTo(out);
-                                            linesCopied.set(readLines(in, splitCount, out, !removeTrailingNewlines));
+                                        }
+
+                                        // Process body
+                                        while (true) {
+                                            final ByteArrayOutputStream dataStream = new ByteArrayOutputStream();
+                                            in.mark(1);
```

Comment:
This "in.mark()" is used to rollback the entire line if its inclusion in the current split exceeds maxFragmentSize. (Reset on line 336.) However, its readlimit should not be 1; it should be the maximum possible length of a line. Since the length of a given line is unknown, recommend using "Integer.MAX_VALUE."